### ISSUE TITLE
write_sav() and datetime gives wrong format

### ISSUE THREAD

**Issue by josmos** [Roche/pyreadstat#69]
According to the Readme *datetime, date, time* should be converted to *numeric with datetime/date/time formatting*.

I am converting strings to datetime64[ns] dtype with:

`df[field_name] = pd.to_datetime(df[field_name], errors="coerce")`

or

`df[field_name] = pd.to_datetime(df[field_name], format='%H:%M', errors="coerce")`

for time, respectively.

The resulting sav variable has *numeric* type with *F8.2* format

How can I convert it to SPSS DATE or TIME format?
Is this a bug? In case I am getting something wrong, please explain how to get the right format.

Thanks
Josef 

Setup Information:

i installed pyreadstat via pipenv

INSTALLED VERSIONS
------------------
commit           : d9fff2792bf16178d4e450fe7384244e50635733
python           : 3.7.5.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-42-generic
Version          : #46~18.04.1-Ubuntu SMP Fri Jul 10 07:21:24 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 1.1.0
numpy            : 1.19.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.1

**Comment by josmos** [comment]
When checking the log files from PSPP when opening files I find error for date and time fields like this one:

`warning: test.sav near offset 0x248: variable V4 with width 0 has invalid print format 0x140800.`
However, if i set variable_display_with I still get the same errors.
Maybe this is causing the incorrect display of datetime values?

**Comment by ofajardo** [comment]
The thing is that all of this works fine when SPSS reads these files. 

Can you produce a file with the features you need in PSPP and put it here in order to examine how the code things?

**Comment by ofajardo** [comment]
In any case the way the variables are written is controlled in the C library underlying pyreadstat, "[Readstat](https://github.com/WizardMac/ReadStat)". Maybe you can write an issue there asking why this is happening. If they would somehow fix it, then I can inherit their changes and it will work for pyreadstat. Otherwise I cannot change those things direclty here in pyreadstat, but I always need it to be fixed in Readstat first.

**Comment by ofajardo** [comment]
OK I used PSPP to write a file with one date variable (with format dd.mm.yy) and one datetime (with format yyyy-mm-dd HH:MM:SS ) (No time variable, because PSPP does not have the option to format as time). When reading in pyreadstat I noticed two interesting things: 

1- Both variables are read as numeric in pyreadstat.
2- The format of the variables is as follows:
{'mydate': 'EDATE8', 'mydatetime': 'YMDHMS20'}
3- Opening the file in SPSS: it did understand the format of mydate, but it did not understand mydatetime and it was shown as numeric.

So, what I think is happening is that PSPP uses its own date and time formats that are not necessarily compatible with SPSS.

Reading PSPP files in pyreadstat is easy to solve for me - I think- , I just need to add things like EDATE8 and YMDHMS20 to the list of recognized formats. I will do so actually. I have seen that PSPP has many possible formats, if somebody would start telling me what those are, then I could add these others as well.

In my opinion since DATE8 and DATETIME8 (the formats that Readstat is using when writing files) are valid SPSS formats, PSPP should support those. I would suggest open an issue on their side.

**Comment by ofajardo** [comment]
I have added EDATE8, EDATE11 (PSPP default) and YMDHMS20 to the list of known formats for SPSS and now pyreadstat can read those formats coming from sav produced by PSPP. Still on dev branch for now.

**Comment by ofajardo** [comment]
Looking at the PSPP source code I found EDATE in several places, among those this comment:

```
/* Parses DATE, ADATE, EDATE, JDATE, SDATE, QYR, MOYR, KWYR,
1120    DATETIME, YMDHMS, MTIME, TIME, and DTIME formats. */
```

so, we are giving TIME8 and it's happy, but for some reason DATE8 and DATETIME8, who theoretically are in the list are not recognized. So maybe they have a bug somewhere else. No idea, you have to ask them.

**Comment by ofajardo** [comment]
the particular log warning that you observed comes from file src/data/sys-file-reader.c, from this function:

```
/* Translates the format spec from sysfile format to internal
1527    format. */
1528 static void
1529 parse_format_spec (struct sfm_reader *r, off_t pos, unsigned int format,
1530                    enum which_format which, struct variable *v,
1531                    int *n_warnings)

1549       if (which == PRINT_FORMAT)
1550         sys_warn (r, pos, _("Variable %s with width %d has invalid print "
1551                             "format 0x%x."),
1552                   var_get_name (v), var_get_width (v), format);
```

No idea what it means or if it is related to the issue here

**Comment by ofajardo** [comment]
the command line tool pspp-dump-sav seems to recognize the formats OK.

```
(base) # pspp-dump-sav export.sav
File header record:
	     Product name: @(#) SPSS DATA FILE - https://github.com/WizardMac/ReadStat
	      Layout code: 2
	       Compressed: 0 (no compression)
	     Weight index: 0
	  Number of cases: 6
	 Compression bias: 100
	    Creation date: 28 Aug 20
	    Creation time: 13:13:26
	       File label: ""
000000b4: variable record #0
	Width: 0 (numeric)
	Variable label: 1
	Missing values code: 0 (no missing values)
	Print format: 160800 (DATETIME8.0)
	Write format: 160800 (DATETIME8.0)
	Name: DATES
	000000d0 Variable label: "mydate"
000000e0: variable record #1
	Width: 0 (numeric)
	Variable label: 1
	Missing values code: 0 (no missing values)
	Print format: 160800 (DATETIME8.0)
	Write format: 160800 (DATETIME8.0)
	Name: TIMES
	000000fc Variable label: "mytime"
```

But trying to read the file in PSPP command line fails:

```
PSPP> get file='export.sav' .

warning: `export.sav' near offset 0xc0: Variable DATES with width 0 has invalid print format 0x160800.
warning: `export.sav' near offset 0xc4: Variable DATES with width 0 has invalid write format 0x160800.
warning: `export.sav' near offset 0xec: Variable TIMES with width 0 has invalid print format 0x160800.
warning: `export.sav' near offset 0xf0: Variable TIMES with width 0 has invalid write format 0x160800.
```

Interestingly the dump command says `Print format: 160800 (DATETIME8.0)` but later PSPP `invalid print format 0x160800`. It truly looks like a bug in PSPP

**Comment by ofajardo** [comment]
Readstat is following the file specification given by PSPP, DATE and DATETIME are specifiied:

https://www.gnu.org/software/pspp/pspp-dev/html_node/Variable-Record.html

**Comment by ofajardo** [comment]
@josmos solved with a big help from @evanmiller

on the latest code on dev, this is working well both in SPSS and PSPP

```
import pandas as pd
import pyreadstat

df = pd.DataFrame({
    "dates": {
        "10001": "nk.nk.2001",
        "10002": "nk.May.2003",
        "10003": "",
        "10004": "nk.nk.2018",
        "10005": None,
        "10006": "30.Dec.2010",
    },
    "times": {
        "10001": "",
        "10002": None,
        "10003": "16:30",
        "10004": "na:na",
        "10005": "10:10",
        "10006": "10:12",
    }
})

df["dates"] = df["dates"].str.replace("nk.nk", "15.Jun")
df["dates"] = df["dates"].str.replace("nk", "15")
df["dates"] = pd.to_datetime(df["dates"], errors="coerce")

df["times"] = pd.to_datetime(df["times"], format='%H:%M', errors="coerce")

df["datetimes"] = df["dates"]
df.dates = df.dates.dt.date
df.times = df.times.dt.time

pyreadstat.write_sav(df, "export.sav",
                     column_labels=["mydate", "mytime", "mydatetime"],
                     variable_measure={"dates": "scale", "times": "scale"},
                     )
```

**Comment by ofajardo** [comment]
fixed in version 1.0.2
